### DISCUSSION TITLE
Is there an option to remove shadow from scanned documents

### DISCUSSION THREAD

**Discussion by Sicos1977** [ImageMagick/ImageMagick#4145] [Help] [answered]
Since I'm a newbie with ImageMagick, I have this question. Is there an option to remove shadow from a scanned document.

See the image below (I cant show the complete image because there is personal information on it). When I use OTSU thresholding on this then to lower shadow part turns black so I was wondering if there is an option to remove the shadow before using OTSU?

![image](https://user-images.githubusercontent.com/6692947/131828504-61b59e6a-25f8-405c-9366-6158baeccabe.png)

**Comment by fmw42** [comment, marked as answer]
Use the Magick.Net equivalent of -lat

`convert 131828504-61b59e6a-25f8-405c-9366-6158baeccabe.png -negate -lat 20x20+10% -negate result.png`

![result](https://user-images.githubusercontent.com/13458218/131869836-3819431f-07e5-417c-918b-b740f87c978f.png)

Or you can get somewhat better results using https://github.com/dlemstra/ImageMagick.FredsScripts.NET textcleaner script
using -f 20 -o 10 options

**Reply by dlemstra** [reply]
Be aware that FredsScript option requires a commercial license for the work that @fmw42  put into this.

**Reply by fmw42** [reply]
License only for commercial use. Non-commercial use is free. Commercial means if you use for profit.

**Reply by dlemstra** [reply]
Looking at their documents I suspect this is for something commercial 😉It's written in Dutch so I can understand what it is about.

**Reply by Sicos1977** [reply]
If I would want to have a license I could not do it anyway since I'm doing development on a Windows 2019 server and there is no bash overthere

**Reply by fmw42** [reply]
You can install Cygwin if you really want. Otherwise, just use -lat equivalent along with -negate equivalent.

**Comment by Sicos1977** [comment]
Dirk you are right that it is for something commercial but I won't make any money on it. I just get paid my salary to do my work and that is in this case get the best OCR result that is possible :-)

**Reply by fmw42** [reply]
If your company make money using the script, then it is for commercial purposes and would need a license. However, you can try for free, until you decide you want to use in production.

**Comment by Sicos1977** [comment]
Is there also a way that you can detect if a document has shadow? Or is the above command fast enough and doesn't it give any speed impact on document without shadow?

**Reply by fmw42** [reply]
Only way I can think is to average the image down to one column. Then look for large consistent changes in intensity along the column. But that means you need to use some scripting to look for the change. The average can be done in ImageMagick.

**Comment by Sicos1977** [comment]
For everyone who ends up here and uses ImageMagick.Net... the anwer from @fmw42  is this translated to ImageMagick.Net

```c#
        private void RemoveShadow(MagickImage image)
        {
            // convert 131828504-61b59e6a-25f8-405c-9366-6158baeccabe.png -negate -lat 20x20+10% -negate result.png
            image.Negate();
            image.AdaptiveThreshold(20, 20, new Percentage(10));
            image.Negate();
        }
```

**Comment by Sicos1977** [comment]
I sometimes have that the method removes to much information and no usefull information is left in the document. I now detect it like this.

```c#
                // convert 131828504-61b59e6a-25f8-405c-9366-6158baeccabe.png -negate -lat 20x20+10% -negate result.png
                WriteToLog($"Removing shadow from image because the option '{Options.Image.RemoveShadow.FullName()}' has been set");
                WriteToLog("Cloning image so that we can undo when to much shadow is removed from the image");
                var beforeRemovingShadow = (MagickImage) Image.Clone(Image.Width, Image.Height);
                var beforeRemovingShadowSize = int.Parse(Image.FormatExpression("%b").Replace("B", string.Empty));
                WriteToLog("Negate image");
                Image.Negate();
                WriteToLog("Adaptive threshold image with 20, 20 + 10%");
                Image.AdaptiveThreshold(20, 20, new Percentage(10));
                WriteToLog("Negate image");
                Image.Negate();
                var afterRemovingShadowSize = Image.ToByteArray(MagickFormat.Png).Length;
                WriteToLog("Shadow removed");

```

Is there a smarter and faster way to clone and compare sizes so I can rollback an image to it's previous state?

**Reply by fmw42** [reply]
I not that the values 20x20+10% might need to be adjusted for different images. They depend upon the size of the text and the degree of noise or variation.
